src/Features/LanguageServer/Protocol/Handler/Formatting/FormatDocumentOnTypeHandler.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [Shared]
+    [ExportLspMethod(Methods.TextDocumentOnTypeFormattingName)]
+    internal class FormatDocumentOnTypeHandler : IRequestHandler<DocumentOnTypeFormattingParams, TextEdit[]>
+    {
+        public async Task<TextEdit[]> HandleRequestAsync(Solution solution, DocumentOnTypeFormattingParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            var edits = new ArrayBuilder<TextEdit>();
+            var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+            if (document != null)
+            {
+                var formattingService = document.Project.LanguageServices.GetService<IEditorFormattingService>();
+                var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+                if (string.IsNullOrEmpty(request.Character))
+                {
+                    return edits.ToArrayAndFree();
+                }
+
+                IList<TextChange> textChanges;
+                // Formatting on new lines must be handled differently.
```

Comment:
>Formatting on new lines must be handled differently. [](start = 19, length = 52)

Perhaps explain why.

Reply:
This is just calling out that Roslyn has a different method for handling formatting on new lines, ```GetFormattingChangesOnReturnAsync``` vs ```GetFormattingChangesAsync```.  I'll remove the comment since it doesn't seem that useful.